akka-actor/src/main/scala/akka/routing/DefaultMetricsBasedResizer.scala

```diff
@@ -0,0 +1,221 @@
+    val (newPerfLog, newRecord) = updatedStats(currentRoutees, messageCounter)
+
+    performanceLog = newPerfLog
+    record = newRecord
+  }
+
+  private[routing] def updatedStats(currentRoutees: immutable.IndexedSeq[Routee], messageCounter: Long): (PerformanceLog, ResizeRecord) = {
+    val now = LocalDateTime.now
+    val currentSize = currentRoutees.length
+
+    val messagesInRoutees = currentRoutees map {
+      case ActorRefRoutee(a: ActorRefWithCell) ⇒
+        a.underlying match {
+          case cell: ActorCell ⇒
+            cell.mailbox.numberOfMessages + (if (cell.currentMessage != null) 1 else 0)
+          case cell ⇒ cell.numberOfMessages
+        }
+      case x ⇒ 0
+    }
+
+    val totalQueueLength = messagesInRoutees.sum
+    val utilized = messagesInRoutees.count(_ > 0)
+
+    val fullyUtilized = utilized == currentSize
+
+    val newUnderutilizationStreak = if (fullyUtilized)
+      None
+    else
+      Some(UnderUtilizationStreak(
+        record.underutilizationStreak.fold(now)(_.start),
+        Math.max(record.underutilizationStreak.fold(0)(_.highestUtilization), utilized)))
+
+    val newPerformanceLog: PerformanceLog =
+      if (fullyUtilized && record.underutilizationStreak.isEmpty && record.checkTime.isDefined) {
+        val totalMessageReceived = messageCounter - record.messageCount
+        val queueSizeChange = record.totalQueueLength - totalQueueLength
+        val totalProcessed = queueSizeChange + totalMessageReceived
+        if (totalProcessed > 0) {
+          val duration = Duration.between(record.checkTime.get, now)
+          val speed = duration dividedBy totalProcessed
+          //exponentially decrease the weight of old speed metrics data
+          val toUpdate = performanceLog.get(currentSize).fold(speed) { oldSpeed ⇒
+            val nanos = (oldSpeed.toNanos * (1d - weightOfLatestMetric)) + (speed.toNanos * weightOfLatestMetric)
+            Duration.ofNanos(nanos.toLong)
+          }
+          performanceLog + (currentSize → toUpdate)
+        } else performanceLog
+      } else performanceLog
+
+    val newRecord = record.copy(
+      underutilizationStreak = newUnderutilizationStreak,
+      messageCount = messageCounter,
+      totalQueueLength = totalQueueLength,
+      checkTime = Some(now))
+
+    (newPerformanceLog, newRecord)
+
+  }
+
+  def resize(currentRoutees: immutable.IndexedSeq[Routee]): Int = {
+    val currentSize = currentRoutees.length
+    val proposedChange =
+      if (record.underutilizationStreak.fold(false)(_.start.isBefore(LocalDateTime.now.minus(downsizeAfterUnderutilizedFor)))) {
+        val downsizeTo = (record.underutilizationStreak.get.highestUtilization * downsizeRatio).toInt
+        Math.min(downsizeTo - currentSize, 0)
+      } else if (performanceLog.isEmpty || record.underutilizationStreak.isDefined) {
+        0
+      } else {
+        if (Random.nextDouble() < explorationProbability)
```

Comment:
please use a `ThreadLocalRandom` (put it in a field), and bound the upper limit of `nextDouble` explicitly